### DISCUSSION TITLE
WASM: directory created using CreateDirectory does not exist after restart

### DISCUSSION THREAD

**Discussion by stefanolson** [unoplatform/uno#4987] [Q&A] [answered]
## Current behavior

WASM: Directory.CreateDirectory creates the directory but only for the current session. When you restart, the directory does not exist. Using CreateFolderAsync works but that is then incompatible with my existing WPF code.

## Expected behavior

Should create directory and when I load the application again the directory should exist.  Not entirely clear how to verify the creation of the directory through the Google debugging tools though.

## How to reproduce it (as minimally and precisely as possible)

run the enclosed project in wasm mode:
[UnoApp.zip](https://github.com/unoplatform/uno/files/5840865/UnoApp.zip)

Click open message box.
it will create a folder using the Windows storage API. 
It will then create a folder using the system.io API.

restart the application. Click open message box.
it will say the folder created using the window storage API exists
it will say the folder created using the system.io API does not exist.

## Workaround

been unable to find anything in the documentation around system.IO or createdirectory.  It will be very difficult to rewrite my WPF code and creates a mess of file system code if I have have different file system code between WPF and Wasm. this code appears to work on android.

## Environment

Nuget Package:

- [ ] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
- [ ] Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia
- [ ] Uno.SourceGenerationTasks
- [ ] Uno.UI.RemoteControl / Uno.WinUI.RemoteControl
- [ ] Other: 

Nuget Package Version(s):

Affected platform(s):

- [ ] iOS
- [ ] Android
- [X] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

IDE:

- [ ] Visual Studio 2017 (version: )
- [X] Visual Studio 2019 (version: 16.8.3)
- [ ] Visual Studio for Mac (version: )
- [ ] Rider Windows (version: )
- [ ] Rider macOS (version: )
- [ ] Visual Studio Code (version: )

Relevant plugins:

- [ ] Resharper (version: )

**Comment by jeromelaban** [comment]
Thanks for the report. 

This is not a bug, but a restriction of the behavior of browsers, where the availability of persisted folders can only be ensured after the return of any async method of `Windows.Storage`.

You should perform the initialization of your application in an async method invoked from `Application.OnLaunched`.

You can view [additional information here](https://platform.uno/docs/articles/features/windows-storage.html#webassembly-support).

**Comment by stefanolson** [comment]
That doesn't actually address the  issue. in the sample code I call a async method of Windows storage and then the system file apis still don't work. documentation does talk about using system I/O file APIs. Is no indication that I'vve done anything wrong with what I'm doing - I'm calling an aysnc method and then calling system.io Apis

**Comment by jeromelaban** [comment]
Thanks for the update.

**Comment by MartinZikmund** [comment, marked as answer]
@stefanolson I have tested the code and found the reason why the folder disappears. In WASM, folders that are persistent across launches need to be specifically marked. This is the case for folders under `ApplicationData` - for example `LocalFolder` and `RoamingFolder` (see [here ](https://github.com/unoplatform/uno/blob/9e291a919a8a7ede90324b37b0dfde9b805b6470/src/Uno.UWP/Storage/ApplicationData.wasm.cs#L19) for the code). This is the reason the folder created using the `StorageFolder` API persists - because it is under the `ApplicationData.LocalFolder.Current` path.

To allow your System.IO folder to persist as well, you need to create it under the same path:

```
var path = Path.Combine(ApplicationData.Current.LocalFolder.Path, systemIoFolderName);
```

**Comment by stefanolson** [comment]
Great. That now works in my sample app.  Hopefully that can be somehow integrated into documentation. I did search on system.io and couldn't find anything. Thanks.

**Comment by jeromelaban** [comment]
@stefanolson that particular bit of information is mentioned [in this document](https://platform.uno/docs/articles/features/windows-storage.html#webassembly-support).
